doc/adr/0002-steppable-cek-machine.md

```diff
@@ -0,0 +1,235 @@
+
+Proposed
+
+## Context
+
+In order to have a minimal viable product of a debugger for Plutus, we need a CEK machine that will give us more information for debugging than our current one.
+
+Implementing the steppable CEK machine is a non-trivial task and involves some design decisions:
+
+- should the debugging machine be a separate one?
+- should we modify the existing CEK machine to support stepping, or implement something entirely new?
+- how do we ensure performance and conformance?
+
+## Decision: A separate machine
+
+We came to the conclusion that the debugging machine should be a separate one. The main reason is that we don't want to slow down the production machine.
+
+## Argument: A separate machine
+
+The debugging machine will give the most accurate information if it was the one in production. However, the drawback of that is that we have to care about its performance.
+
+We only care about performance if the steppable CEK machine will *replace* the production machine instead of being a separate one.
+
+Our current CEK machine's performance is sensitive to changes. For example, exporting the `computeCek` function from the module causes the CEK machine to become slower by up to 25%. Refactoring the machine in preparation for the stepper also shows significant slow downs. See [PR#4909](https://github.com/input-output-hk/plutus/pull/4909). We have enough evidence to show that replacing the machine without slow downs is impossible.
+
+We consciously chose to keep up the production CEK machine's performance by adding a separate machine for debugging. We will mitigate the drawback of this by keeping the implementation as close to the production machine as possible.
+
+## Alternative: polymorphic compute/return steps
+
+This approach has been suggested to potentially not introduce slow downs while not having a separate debugging CEK machine.
+
+As long as the debugging machine has the same type, we can alter `computeCek`/`returnCek` to be polymorphic over a type-level `Bool` specifying if we’re in debug mode or not. Then we demote it to the term level in the definition of `computeCek`/`returnCek` and branch on the `Bool` thus implementing different logic depending on whether we're in debug mode or not. This promotion to the type level allows us to statically instantiate the `Bool` in an instance and thus make GHC compile the whole worker of the CEK machine twice: once in debug mode and once in production mode. Theoretically, GHC will take care to remove all the dead debug code when in production mode.
+
+We currently decide to not take this approach because:
+
+- It's not clear whether we can implement it this way. Implementing a separate machine is simpler.
+- Some slowdown may still happen due to GHC optimizations being very unreliable.
+- It's difficult to detect slow downs due to this as we add more builtin functions.
+
+If the separate machine turns out to not provide correct debugging information, we may revisit this option again.
```

Comment:
Which is why we should generate traces in the debugging machine and in the production one (by piggybacking on the costing machinery) and test that those are the same.

Reply:
Why would it not provide correct debugging information? I'm confused by the implication that they would behave differently for some reason?

Reply:
The point is that they shouldn't! And we need to make sure of that. See my [other comment](https://github.com/input-output-hk/plutus/pull/4915#discussion_r1001934889).

Reply:
The sentence didn't read to me like that. It read to me like "if we implement it correctly and somehow it gives wrong information, then we should...", rather than "if we struggle to implement the separate machine correctly, then we should..."